pkg/controller/velero/s3.go

```diff
@@ -0,0 +1,125 @@
+package velero
+
+import (
+	"fmt"
+	"time"
+
+	veleroCR "github.com/openshift/managed-velero-operator/pkg/apis/managed/v1alpha1"
+	"github.com/openshift/managed-velero-operator/pkg/s3"
+
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	awss3 "github.com/aws/aws-sdk-go/service/s3"
+	"github.com/google/uuid"
+)
+
+const (
+	bucketPrefix = "managed-velero-backups-"
+)
+
+func (r *ReconcileVelero) provisionS3(reqLogger logr.Logger, s3Client *awss3.S3, instance *veleroCR.Velero) (reconcile.Result, error) {
+	var err error
+	bucketLog := reqLogger.WithValues("S3Bucket.Name", instance.Status.S3Bucket.Name, "S3Bucket.Region", s3Client.Client.Config.Region)
+
+	// This switch handles the provisioning steps/checks
+	switch {
+
+	// We don't yet have a bucket name selected
+	case instance.Status.S3Bucket.Name == "":
+		log.Info("No S3 bucket defined")
+		proposedName := generateBucketName(bucketPrefix)
+		proposedBucketExists, err := s3.DoesBucketExist(s3Client, proposedName)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if proposedBucketExists {
+			return reconcile.Result{}, fmt.Errorf("proposed bucket %s already exists, retrying", proposedName)
+		}
+
+		log.Info("Setting proposed bucket name", "S3Bucket.Name", proposedName)
+		instance.Status.S3Bucket.Name = proposedName
+		instance.Status.S3Bucket.Provisioned = false
+		return reconcile.Result{}, r.statusUpdate(reqLogger, instance)
+
+	// We have a bucket name, but haven't kicked off provisioning of the bucket yet
+	case instance.Status.S3Bucket.Name != "" && !instance.Status.S3Bucket.Provisioned:
+		bucketLog.Info("S3 bucket defined, but not provisioned")
+
+		// Create S3 bucket
+		bucketLog.Info("Creating S3 Bucket")
+		err = s3.CreateBucket(s3Client, instance.Status.S3Bucket.Name)
+		if err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				switch aerr.Code() {
+				case awss3.ErrCodeBucketAlreadyExists:
+					bucketLog.Info("Bucket exists, but is not owned by current user; retrying")
```

Comment:
Does it leak information to emit who the current user is?

Reply:
No.. this is basically the "name collision" check, as S3 bucket names are globally unique. We have no clue who actually owns it, only that it's not available. The chance of hitting this is rare (we have a named prefix for the operator + a UUID), but still wanted to handle the condition. We just pick a new name and retry.

Reply:
Sounds good.